playground/backend/cmd/server/controller.go

```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
+
+	// set to cache pipelineId: cache.SubKey_Status: Status_STATUS_TIMEOUT
+	err := cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
+
+// cleanUp removes all prepared folders for received LifeCycle
+func cleanUp(lc *fs_tool.LifeCycle) {
+	log.Println("parallelRunCode complete")
+
+	log.Println("parallelRunCode: DeleteFolders ...")
+	err := lc.DeleteFolders()
+	if err != nil {
+		log.Printf("parallelRunCode: DeleteFolders: " + err.Error())
+	}
+	log.Println("parallelRunCode: DeleteFolders complete")
+}
+
+// processErrDuringValidate processes error received during Validate step
+func processErrDuringValidate(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: Validate: " + err.Error())
+
+	// set to cache pipelineId: cache.SubKey_Status: pb.Status_STATUS_ERROR
+	err = cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
+
+// processErrDuringCompile processes error received during Compile step
+func processErrDuringCompile(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: Compile: " + err.Error())
+
+	// set to cache pipelineId: cache.SubKey_Status: pb.Status_STATUS_ERROR
+	err = cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+
+	// set to cache pipelineId: cache.SubKey_CompileOutput: err.Error()
+	err = cacheService.SetValue(pipelineId, cache.SubKey_CompileOutput, err.Error())
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
+
+// processErrDuringGetExecutableName processes error received during getting executable file name
+func processErrDuringGetExecutableName(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: get executable file name: " + err.Error())
+
+	// set to cache pipelineId: cache.SubKey_Status: pb.Status_STATUS_ERROR
+	err = cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
+
+// processErrDuringRun processes error received during Run step
+func processErrDuringRun(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: Run: " + err.Error())
+
+	// set to cache pipelineId: cache.SubKey_RunOutput: err.Error()
+	err = cacheService.SetValue(pipelineId, cache.Subkey_RunOutput, err.Error())
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+
+	// set to cache pipelineId: cache.SubKey_Status: pb.Status_STATUS_ERROR
+	err = cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
```

Comment:
I see too many common actions in these methods – can we create a generic function for such processing?